common/cluster/metadata.go

```diff
@@ -131,76 +163,155 @@ func NewMetadata(
+	return (version1-version2)%m.failoverVersionIncrement == 0
 }
 
-func (metadata *metadataImpl) IsMasterCluster() bool {
-	return metadata.masterClusterName == metadata.currentClusterName
+func (m *metadataImpl) IsMasterCluster() bool {
+	return m.masterClusterName == m.currentClusterName
 }
 
 // GetMasterClusterName return the master cluster name
-func (metadata *metadataImpl) GetMasterClusterName() string {
-	return metadata.masterClusterName
+func (m *metadataImpl) GetMasterClusterName() string {
+	return m.masterClusterName
 }
 
 // GetCurrentClusterName return the current cluster name
-func (metadata *metadataImpl) GetCurrentClusterName() string {
-	return metadata.currentClusterName
+func (m *metadataImpl) GetCurrentClusterName() string {
+	return m.currentClusterName
 }
 
 // GetAllClusterInfo return the all cluster name -> corresponding information
-func (metadata *metadataImpl) GetAllClusterInfo() map[string]config.ClusterInformation {
-	return metadata.clusterInfo
+func (m *metadataImpl) GetAllClusterInfo() map[string]config.ClusterInformation {
+	return m.clusterInfo
 }
 
 // ClusterNameForFailoverVersion return the corresponding cluster name for a given failover version
-func (metadata *metadataImpl) ClusterNameForFailoverVersion(failoverVersion int64) string {
+func (m *metadataImpl) ClusterNameForFailoverVersion(failoverVersion int64) string {
 	if failoverVersion == common.EmptyVersion {
-		return metadata.currentClusterName
+		return m.currentClusterName
 	}
 
-	initialFailoverVersion := failoverVersion % metadata.failoverVersionIncrement
+	initialFailoverVersion := failoverVersion % m.failoverVersionIncrement
 	// Failover version starts with 1.  Zero is an invalid value for failover version
 	if initialFailoverVersion == common.EmptyVersion {
-		initialFailoverVersion = metadata.failoverVersionIncrement
+		initialFailoverVersion = m.failoverVersionIncrement
 	}
 
-	clusterName, ok := metadata.versionToClusterName[initialFailoverVersion]
+	clusterName, ok := m.versionToClusterName[initialFailoverVersion]
 	if !ok {
 		panic(fmt.Sprintf(
 			"Unknown initial failover version %v with given cluster initial failover version map: %v and failover version increment %v.",
 			initialFailoverVersion,
-			metadata.clusterInfo,
-			metadata.failoverVersionIncrement,
+			m.clusterInfo,
+			m.failoverVersionIncrement,
 		))
 	}
 	return clusterName
 }
+
+// GetSearchAttributes returns all search attributes (including system and build-in) for specified index.
+func (m *metadataImpl) GetSearchAttributes(indexName string, bypassCache bool) (map[string]enumspb.IndexedValueType, error) {
+	refreshCache := func(bypassCache bool, saCache searchAttributesCache, now time.Time) bool {
+		return bypassCache ||
+			saCache.lastRefresh.Add(searchAttributeCacheRefreshInterval).Before(now) ||
+			saCache.searchAttributes == nil
+	}
+
+	now := m.timeSource.Now()
+	saCache := m.searchAttributesCache.Load().(searchAttributesCache)
+
```

Comment:
`GetSearchAttributes` and `SaveSearchAttributes` APIs are the core of this PR. They are not used yet but will replace dynamic config in upcoming PR.

It uses read optimized cache: `atomic` for read and `Mutex` for update.